File: crates/posthaste-authority-server/src/mutations/automation.rs
use super::*;

impl AccountMutationService {
    pub fn preview_automation_rule(
        &self,
        request: AutomationRulePreviewMutation,
    ) -> Result<AutomationRulePreviewResult, RuntimeError> {
        let (_, total) = self.service.count_messages_by_rule(&request.condition)?;
        let page = self.service.query_message_page_by_rule(
            &request.condition,
            request.limit,
            None,
            MessageSortField::Date,
            SortDirection::Desc,
        )?;
        Ok(AutomationRulePreviewResult {
            total,
            items: page.items,
        })
    }
}

pub(super) fn normalize_automation_rules(rules: &[AutomationRule]) -> Vec<AutomationRule> {
    rules
        .iter()
        .map(|rule| AutomationRule {
            id: rule.id.trim().to_string(),
            name: rule.name.trim().to_string(),
            enabled: rule.enabled,
            triggers: rule.triggers.clone(),
            condition: rule.condition.clone(),
            actions: rule
                .actions
                .iter()
                .map(normalize_automation_action)
                .collect(),
            backfill: rule.backfill,
        })
        .collect()
}

fn normalize_automation_action(action: &AutomationAction) -> AutomationAction {
    match action {
        AutomationAction::ApplyTag { tag } => AutomationAction::ApplyTag {
            tag: tag.trim().to_string(),
        },
        AutomationAction::RemoveTag { tag } => AutomationAction::RemoveTag {
            tag: tag.trim().to_string(),
        },
        AutomationAction::MarkRead => AutomationAction::MarkRead,
        AutomationAction::MarkUnread => AutomationAction::MarkUnread,
        AutomationAction::Flag => AutomationAction::Flag,
        AutomationAction::Unflag => AutomationAction::Unflag,
        AutomationAction::MoveToMailbox { mailbox_id } => AutomationAction::MoveToMailbox {
            mailbox_id: MailboxId::from(mailbox_id.as_str().trim()),
        },
    }
}

File: apps/client/frontend/src/data/transport/api/mail/automation.ts
import type { MessageSummary } from './mail'
import type { MailQueryRule } from './mailQuery'

export type AutomationTrigger = 'messageArrived' | 'messageChanged' | 'manual'

export type AutomationAction =
  | { kind: 'applyTag'; tag: string }
  | { kind: 'removeTag'; tag: string }
  | { kind: 'markRead' }
  | { kind: 'markUnread' }
  | { kind: 'flag' }
  | { kind: 'unflag' }
  | { kind: 'moveToMailbox'; mailboxId: string }

export interface AutomationRule {
  id: string
  name: string
  enabled: boolean
  triggers: AutomationTrigger[]
  condition: MailQueryRule
  actions: AutomationAction[]
  backfill: boolean
}

export interface AutomationRulePreviewInput {
  condition: MailQueryRule
  limit?: number
}

export interface AutomationRulePreviewResponse {
  total: number
  items: MessageSummary[]
}
